test/Parse/diagnose_availability.swift

```diff
@@ -0,0 +1,27 @@
+// RUN: %target-typecheck-verify-swift 
+
+// SR-4231: Misleading/wrong error message for malformed @available
+
+@available(OSX 10.6, *) // no error
+func availableSince10_6() {}
+
+@available(OSX, introduced: 10.0, deprecated: 10.12) // no error
+func introducedFollowedByDeprecated() {}
+
+@available(OSX 10.0, deprecated: 10.12)
```

Comment:
I think that it probably could be made to recover, but I don't know how to do it. (I’m assuming that recovery means being able to emit the right attribute and continue parsing.)

As far as I understand how this works `parseAvailabilitySpecList` (this code) parses a comma separated list of short for availability specifications, and returns it to the `DAK_Available` case of `parseNewDeclAttribute` which turns each specification into their full attribute.

I think that in order to recover, it would need to pass some additional information back so that an `AvailableAttr` could be created with both an introduced and deprecated parameter, or so that the current token could be parsed differently. In either case, I need to consume tokens until the closing parenthesis.

Not that it would be a lot of work, but I don’t know if it would add much on top of the diagnostic and fix-it. That said; if you think it's worth doing, I would see it as a good learning opportunity.